src/Bicep.Core/Registry/IModuleRegistry.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Modules;
+using System;
+using System.Collections.Generic;
+
+namespace Bicep.Core.Registry
+{
+    /// <summary>
+    /// An implementation of a Bicep module registry.
+    /// </summary>
+    public interface IModuleRegistry
+    {
+        /// <summary>
+        /// Gets the scheme used by this registry in module references.
+        /// </summary>
+        string Scheme { get; }
+
+        /// <summary>
+        /// Attempts to parse the specified unqualified reference or returns a failure builder.
+        /// </summary>
+        /// <param name="reference">The unqualified module reference</param>
+        /// <param name="failureBuilder">set to an error builder if parsing fails when null is returned</param>
+        ModuleReference? TryParseModuleReference(string reference, out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder);
+
+        /// <summary>
+        /// Returns true if the specified module is already cached in the local cache. 
+        /// </summary>
+        /// <param name="reference">The reference to the module.</param>
+        bool IsModuleRestoreRequired(ModuleReference reference);
+
+        /// <summary>
+        /// Returns a URI to the entry point module.
+        /// </summary>
+        /// <param name="parentModuleUri">The parent URI</param>
+        /// <param name="reference">The module reference</param>
+        /// <param name="failureBuilder">set to an error builder if parsing fails when null is returned</param>
+        /// <returns></returns>
+        Uri? TryGetLocalModuleEntryPointPath(Uri parentModuleUri, ModuleReference reference, out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder);
```

Comment:
>Uri parentModuleUri [](http://example.com/codeflow?start=45&length=19)

This is a weak part of this design. Ideally, this info would be captured inside `LocalModuleReference` objects but that would mean a `ModuleDeclarationSyntax` would need to know this as well.

Thoughts?